pkg/signal/trap_linux_test.go

```diff
@@ -0,0 +1,99 @@
+// +build linux
+
+package signal
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestTrap(t *testing.T) {
+	sigmap := map[string]os.Signal{
+		"TERM": syscall.SIGTERM,
+		"QUIT": syscall.SIGQUIT,
+		"INT":  os.Interrupt,
+	}
+
+	if os.Getenv("TEST_TRAP") == "1" {
+		defer time.Sleep(5 * time.Second)
+
+		Trap(func() {
+			time.Sleep(1 * time.Second)
+			os.Exit(99)
+		})
+
+		go func() {
+			p, err := os.FindProcess(os.Getpid())
+			if err != nil {
+				t.Fatalf("unable to get current process id: %v", err)
+			}
+
+			switch s := os.Getenv("SIGNAL_TYPE"); s {
+			case "TERM":
+				for {
+					p.Signal(sigmap[s])
+				}
+			case "QUIT":
+				p.Signal(sigmap[s])
+			case "INT":
+				p.Signal(sigmap[s])
+			}
+		}()
+		select {}
+	}
+
+	for k, v := range sigmap {
+		cmd := exec.Command(os.Args[0], "-test.run=TestTrap")
+		cmd.Env = append(os.Environ(), "TEST_TRAP=1", fmt.Sprintf("SIGNAL_TYPE=%s", k))
```

Comment:
Instead of re-execing this process with a special environment what do you think about having a separate test binary and running that?

The source could go into `./pkg/signal/testfiles/main.go`, and this test can `go run ...` to run the process.